scripts/solend_obligation_fetcher.py

```diff
@@ -0,0 +1,301 @@
+"""
+Script for Solend loan states collection.
+"""
+# Standard library imports
+import logging
+import os
+import struct
+import time
+from typing import Any, Dict, List, Tuple
+
+# Related third-party imports
+import base58
+import borsh_construct as borsh
+import pandas as pd
+import requests
+from anchorpy.borsh_extension import BorshPubkey
+from solana.exceptions import SolanaRpcException
+from solana.rpc.api import Client
+from solana.rpc.types import MemcmpOpts
+from solders.pubkey import Pubkey
+from solders.rpc.responses import GetProgramAccountsResp
+
+# Local application/library specific imports
+import db
+from src.loans.loan_state import store_loan_states
+from src.loans.solend import SolendState
+
+# logger
+LOGGER = logging.getLogger(__name__)
+
+# Constants
+SOLEND_PROGRAM_ID = Pubkey.from_string("So1endDq2YkqhipRh3WViPa8hdiSpxWy6z3Z6tMCpAo")
+LENDING_MARKET_MAIN = Pubkey.from_string("4UpD2fh7xH3VP9QQaXtsS1YY3bxzWhtfpks7FatyKvdY")
+OBLIGATION_LEN = 1300
+AUTHENTICATED_RPC_URL = os.getenv("RPC_URL")
+
+
+LastUpdateLayout = borsh.CStruct(
+    "slot" / borsh.U64,
+    "stale" / borsh.U8
+)
+
+# struct layouts
+ObligationLayout = borsh.CStruct(
+    "version" / borsh.U8,
+    "last_update" / LastUpdateLayout,
+    "lendingMarket" / BorshPubkey,
+    "owner" / BorshPubkey,
+    "depositedValue" / borsh.U128,
+    "borrowedValue" / borsh.U128,
+    "allowedBorrowValue" / borsh.U128,
+    "unhealthyBorrowValue" / borsh.U128,
+    "padding" / borsh.U8[64],
+    "depositsLen" / borsh.U8,
+    "borrowsLen" / borsh.U8,
+    "dataFlat" / borsh.U8[1096],
+)
+
+OBLIGATION_COLLATERAL_LAYOUT = "32s Q 16s 32x"  # 32 bytes PublicKey, uint64, uint128, 32 bytes padding
+OBLIGATION_LIQUIDITY_LAYOUT = "32s 16s 16s 16s 32x"  # 32 bytes PublicKey, three uint128, 32 bytes padding
+
+# isolated pool addresses
+ISOLATED_POOLS = {
+    "TURBO_SOL_POOL": "7RCz8wb6WXxUhAigok9ttgrVgDFFFbibcirECzWSBauM",
+    "INVICTUS_POOL": "5i8SzwX2LjpGUxLZRJ8EiYohpuKgW2FYDFhVjhGj66P1",
+    "STEP_POOL": "DxdnNmdWHcW6RGTYiD5ms5f7LNZBaA7Kd1nMfASnzwdY",
+    "BONFIDA_POOL": "91taAt3bocVZwcChVgZTTaQYt2WpBVE3M9PkWekFQx4J",
+    "STAR_ATLAS_POOL": "99S4iReDsyxKDViKdXQKWDcB6C3waDmfPWWyb5HAbcZF",
+    "DOG_POOL": "HASr6hiYVoRcVXk3GttC4PjBBPQ3sGYDzE7HSPJdcke6",
+    "STABLE_POOL": "GktVYgkstojYd8nVXGXKJHi7SstvgZ6pkQqQhUPD7y7Q",
+    "NFT_POOL": "29yTiqjGdoNiRLMVc7ZoqFpbW3gkmefwMG9SUiMMD4J9",
+    "COIN98_POOL": "7tiNvRHSjYDfc6usrWnSNPyuN68xQfKs1ZG2oqtR5F46",
+    "UXD_POOL": "HCuEqcXaGeioiJf5vNMTyQC7HMPqJm5aZPkSjA2qceDS",
+    "STEPN_POOL": "BjsAGLZzAgBUsiaTTDQv7PWDUDL9dQfKvYwb4q6FoDuD",
+    "SHADOW_POOL": "Foo9vqN6fj1NyymHmD1gwZkgVgEqzNSrwyeqyoLYGe7j"
```

Comment:
Can this config be fetched from somewhere as well? So that we don't have to manually add new pools in the future?

Reply:
i ll put task in backlog to figure it out